### PR TITLE
[jvm-packages] update rabit, surface new changes to spark, add parity and failure tests

### PR DESCRIPTION
**Pull request by chenqin**
This pr doesn't change or cover checkpoint restore behavior.

It allow user set of rabit parameters to xgb4j-spark within configurations. 
By default, those parameters were set consistent with values in rabit project.

### CODE REVIEW

**Comment by chenqin** [comment]
this pr is blocked by bug fix in rabit 
https://github.com/dmlc/rabit/pull/106

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala [hunk 1]
```diff
@@ -77,4 +77,17 @@ class XGBoostConfigureSuite extends FunSuite with PerTest {
     }
     ss.conf.unset("xgboost.spark.ignoreSsl")
   }
+
+  test("Check setting rabit configs no affect on prediction") {
+    val training = buildDataFrame(Classification.train)
+    val testDM = new DMatrix(Classification.test.iterator, null)
+    val paramMap = Map("eta" -> "1", "max_depth" -> "2", "silent" -> "1",
```
Comment (trivialfis) [inline jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala]: Silent parameter is not be used.  Use `verbosity` instead.

**Comment by chenqin** [comment]
@trivialfis do you happen to know how to set default value of raibt option DMLC_ROOT in xgboost Cmake file?
I tried use following way and seems not working, only when I pass with cmake -DDMLC_ROOT=absolute path to dmlc-core .. seems working fine.

```
# rabit
set(DMLC_ROOT ${PROJECT_SOURCE_DIR}/dmlc-core)
add_subdirectory(${PROJECT_SOURCE_DIR}/rabit)
set_target_properties(rabit
  PROPERTIES
  CXX_STANDARD 11
  CXX_STANDARD_REQUIRED ON
  POSITION_INDEPENDENT_CODE ON)
list(APPEND LINKED_LIBRARIES_PRIVATE rabit)
```

**Comment by trivialfis** [comment]
@chenqin I built XGBoost successfully on my local machine in #4880 .

**Comment by chenqin** [comment]
done with cleanup, xgboost now build with rabit

**Comment by hcho3** [comment]
@chenqin Nicely done!

File: .gitmodules [hunk 2]
```diff
@@ -3,7 +3,7 @@
 	url = https://github.com/dmlc/dmlc-core
 [submodule "rabit"]
 	path = rabit
-	url = https://github.com/dmlc/rabit
+	url = https://github.com/chenqin/rabit
```
Comment (CodingCat) [inline .gitmodules]: can we revert this?

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala [hunk 3]
```diff
@@ -296,7 +296,13 @@ class XGBoostClassificationModel private[ml](
           if (batchCnt == 0) {
             val rabitEnv = Array(
               "DMLC_TASK_ID" -> TaskContext.getPartitionId().toString,
-              "DMLC_WORKER_STOP_PROCESS_ON_ERROR" -> "false").toMap
+              "DMLC_WORKER_CONNECT_RETRY" -> String.valueOf(getConnectRetry),
+              "DMLC_WORKER_STOP_PROCESS_ON_ERROR" -> "false",
+              "rabit_bootstrap_cache" -> String.valueOf(getBootstrapCache),
+              "rabit_debug" -> String.valueOf(getRabitDebug),
+              "rabit_reduce_buffer" -> String.valueOf(getReduceBuffer),
+              "rabit_reduce_ring_mincount" -> String.valueOf(getRingReduceMin)
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala]: let's build a common class for both classifier and regressor and put these values there

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala [hunk 4]
```diff
@@ -296,7 +296,13 @@ class XGBoostClassificationModel private[ml](
           if (batchCnt == 0) {
             val rabitEnv = Array(
               "DMLC_TASK_ID" -> TaskContext.getPartitionId().toString,
-              "DMLC_WORKER_STOP_PROCESS_ON_ERROR" -> "false").toMap
+              "DMLC_WORKER_CONNECT_RETRY" -> String.valueOf(getConnectRetry),
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala]: it should be toString in scala

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostRegressor.scala [hunk 5]
```diff
@@ -273,7 +273,13 @@ class XGBoostRegressionModel private[ml] (
           if (batchCnt == 0) {
             val rabitEnv = Array(
               "DMLC_TASK_ID" -> TaskContext.getPartitionId().toString,
-              "DMLC_WORKER_STOP_PROCESS_ON_ERROR" -> "false").toMap
+              "DMLC_WORKER_CONNECT_RETRY" -> String.valueOf(getConnectRetry),
+              "DMLC_WORKER_STOP_PROCESS_ON_ERROR" -> "false",
+              "rabit_bootstrap_cache" -> String.valueOf(getBootstrapCache),
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostRegressor.scala]: it should be toString in scala

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/params/GeneralParams.scala [hunk 6]
```diff
@@ -166,6 +164,42 @@ private[spark] trait GeneralParams extends Params {
...
+
+  final def rabitDebug: IntParam = new IntParam(this, "rabit_debug",
+    "enable more verbose rabit logging to stdout", (debug: Int) => debug == 0 || debug == 1)
+
+  final def getRabitDebug: Int = ${rabitDebug}
+
+  final def connectRetry: IntParam = new IntParam(this, "DMLC_WORKER_CONNECT_RETRY",
+    "number of retry worker do before fail", ParamValidators.gtEq(1))
+
+  final def getConnectRetry: Int = ${connectRetry}
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/params/GeneralParams.scala]: can we have another Params class to save all rabit params (avoid exploiting this class)

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala [hunk 7]
```diff
@@ -77,4 +77,17 @@ class XGBoostConfigureSuite extends FunSuite with PerTest {
     }
     ss.conf.unset("xgboost.spark.ignoreSsl")
   }
+
+  test("Check setting rabit configs no affect on prediction") {
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala]: this test is not enough, we should test all passed in rabit params are taking effect in rabit layer

File: .gitmodules (hunk 2 above)
Comment (hcho3) [reply .gitmodules]: @CodingCat We need to merge https://github.com/dmlc/rabit/pull/106 first.

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala (hunk 7 above)
Comment (chenqin) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala]: I see, then we need new api in rabit and jni to read parameter out. like https://github.com/dmlc/rabit/commit/edac14220669bab9c3605ee1664d3bce816dbf3c

File: .gitmodules (hunk 2 above)
Comment (chenqin) [reply .gitmodules]: will swing back after https://github.com/dmlc/rabit/pull/106 and https://github.com/chenqin/rabit/commit/edac14220669bab9c3605ee1664d3bce816dbf3c merged

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala (hunk 3 above)
Comment (chenqin) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala]: sounds good, will do.

Comment (chenqin) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala]: done

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostRegressor.scala (hunk 5 above)
Comment (chenqin) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostRegressor.scala]: done

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/params/GeneralParams.scala (hunk 6 above)
Comment (chenqin) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/params/GeneralParams.scala]: done

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala (hunk 7 above)
Comment (chenqin) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala]: or we can mock rabit.init

Comment (chenqin) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala]: giving a quick check neither of those options seems easy to add on this pr. I would prefer adding a follow up pr and adding unittest based on mockito to xgboost4j related projects. What do you think @CodingCat

**Comment by chenqin** [comment]
saw connection errors in failed tests, can we restart tests @trivialfis

**Comment by trivialfis** [comment]
Restarted.

**Comment by trivialfis** [comment]
@hcho3 Could you please help taking a look at the GPU instances, seems blocking the tests.

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala (hunk 7 above)
Comment (CodingCat) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala]: (1) the test here is not reasonable, why testing `no affect on prediction` is enough for this?

(2) why writing test is difficult?

(2) even writing test is difficult for whatever reason we should ensure we can write test first (i.e. finish whatever needed to adding test for this PR first), and then add this PR later

**Comment by hcho3** [comment]
@trivialfis It looks like there was a temporary capacity problem in AWS for GPU instance type. I restarted the test.

**Comment by trivialfis** [comment]
@hcho3 Thanks for the information.

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala (hunk 7 above)
Comment (chenqin) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala]: I agree with (3), my concern is really if this diff gets too big to review.
(2) try to support powermock in spark tests, let met do more research on this

Comment (chenqin) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala]: added test verify Rabit get init with expected params

**Comment by chenqin** [comment]
per feedback from @CodingCat, we added tests which verify newly surfaced parameters user passed in get properly populated into Rabit. 

we have tests cover those paramter passed in were parsed and taken into consideration tests https://github.com/dmlc/rabit/blob/master/test/cpp/allreduce_base_test.cpp

we have integration test which covers those parameter actually were taking effects in rabit https://github.com/dmlc/rabit/blob/master/test/test.mk#L15

cc @hcho3

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala [hunk 8]
```diff
@@ -77,4 +80,27 @@ class XGBoostConfigureSuite extends FunSuite with PerTest {
...
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala]: we should also have set of tests in xgb-spark ensuring that, changing rabit params can still make training (approx&hist) run with verified results

Comment (CodingCat) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala]: for this particular test, it is confusing to me that why "prediction" is something error-prone when there is changing of rabit params?

Comment (chenqin) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala]: good call, In fact, we should only check if parameters were propagated into rabit. As stated above, this pr doesn't suppose to change behavior of xgb-spark.

File: jvm-packages/xgboost4j/src/test/java/ml/dmlc/xgboost4j/java/DMatrixTest.java [hunk 9]
```diff
@@ -315,6 +315,7 @@ public void testTrainWithDenseMatrixRef() throws XGBoostError {
         assertEquals(1, preds.length);
         assertArrayEquals(new float[]{(float) (i + 1)}, preds[0], 1e-2f);
       }
+      assertEquals(rabitEnv, Rabit.rabitEnvs);
```
[the remark this answers is not in the document]
Comment (chenqin) [reply jvm-packages/xgboost4j/src/test/java/ml/dmlc/xgboost4j/java/DMatrixTest.java]: removed

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/params/RabitParams.scala [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+  final def connectRetry: IntParam = new IntParam(this, "DMLC_WORKER_CONNECT_RETRY",
+    "number of retry worker do before fail", ParamValidators.gtEq(1))
+
+  final def getConnectRetry: Int = ${connectRetry}
+
+  setDefault(ringReduceMin -> (32 << 10), reduceBuffer -> "256MB", bootstrapCache -> 0,
+    rabitDebug -> 0, connectRetry -> 5)
+
+  def XGBoostToRabitParams(xgboostParams: Map[String, Any]): Unit = {
```
[the remark this answers is not in the document]
Comment (chenqin) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/params/RabitParams.scala]: removed unused code, there are some subtle differences of having captal letter parameters v.s not having, Rabit only has string and int.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 11]
```diff
@@ -221,6 +221,19 @@ private[this] class XGBoostExecutionParamsFactory(rawParams: Map[String, Any], s
     xgbExecParam.setRawParamMap(overridedParams)
     xgbExecParam
   }
+
+  def buildRabitParams: Map[String, String] = Map(
```
[the remark this answers is not in the document]
Comment (chenqin) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: done

**Comment by trivialfis** [comment]
Do you have anything else needs to be merged in rabit first?  Otherwise I would like to upgrade rabit first so I have more time for testing.

**Comment by chenqin** [comment]
> Do you have anything else needs to be merged in rabit first? Otherwise I would like to upgrade rabit first so I have more time for testing.

Should be good, please upgrade rabit and start testing

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala (hunk 7 above)
Comment (chenqin) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala]: moved all rabit tests to a separate suite

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala (hunk 8 above)
Comment (chenqin) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostConfigureSuite.scala]: moved to rabitRegressionSuite

File: .gitmodules (hunk 2 above)
Comment (chenqin) [reply .gitmodules]: and this change https://github.com/dmlc/rabit/pull/120

**Comment by chenqin** [comment]
@CodingCat per feedback, 
- split rabit tests to [separate test suite]( https://github.com/dmlc/xgboost/pull/4876/files#diff-40a909d943bf182d92e0f0241831b727R1)
- verify actual failure handing from in native layer
- verify both classification and regression parity w/o enable new settings

File: jvm-packages/create_jni.py [hunk 12]
```diff
@@ -6,6 +6,7 @@
 import subprocess
 import sys
 from contextlib import contextmanager
+import getopt
```
Comment (trivialfis) [inline jvm-packages/create_jni.py]: Not used?  Argparser is also quite good.

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostRabitRegressionSuite.scala [hunk 13]
```diff
@@ -0,0 +1,109 @@
+/*
+ Copyright (c) 2014 by Contributors
```
Comment (trivialfis) [inline jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostRabitRegressionSuite.scala]: 2014¿

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/params/RabitParams.scala [hunk 14]
```diff
@@ -0,0 +1,62 @@
+/*
+ Copyright (c) 2014 by Contributors
```
Comment (trivialfis) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/params/RabitParams.scala]: 2014¿

File: jvm-packages/create_jni.py (hunk 12 above)
Comment (chenqin) [reply jvm-packages/create_jni.py]: removed for now, we only accept on parameter for testing usage, might be overkill to allow create_jni.py run as commandline tool

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/params/RabitParams.scala (hunk 14 above)
Comment (chenqin) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/params/RabitParams.scala]: updated scala stype check

**Comment by chenqin** [comment]
@trivialfis looks like ci build were stoped for unclear reason, can you help restart tests.

**Comment by trivialfis** [comment]
Done.  Allocating GPU doesn't seem breezy.

**Comment by chenqin** [comment]
humm, also ci https://travis-ci.org/chenqin/xgboost/builds/598803845

**Comment by chenqin** [comment]
Per offline convo with @CodingCat , we might just review/merge this pr even GPU tests were not able to run.  @trivialfis

**Comment by trivialfis** [comment]
@hcho3 Could you please help confirming that there's nothing we can do at the moment?

**Comment by hcho3** [comment]
@trivialfis I have no idea what happened. CloudTrail doesn't show any failure due to global capacity. I'm starting to have less trust in the EC2 plugin Jenkins. It may be better off if we wrote a custom autoscaling logic in AWS Lambda instead, something like https://github.com/apache/incubator-mxnet-ci/tree/master/services/jenkins-autoscaling

**Comment by chenqin** [comment]
per offline convo from @CodingCat, updated regression parity tests w/o rabit   
```assert(math.abs(p1 - p2) < 0.00001f)```, also enable rabit assert error propagated to xgboost4j by default. All tests passed.

**Comment by chenqin** [comment]
@trivialfis please help review and merge if seems okay.

**Comment by trivialfis** [comment]
Epic.  :-)

**Comment by CodingCat** [comment]
I will review in the weekend, please don't rush

**Comment by trivialfis** [comment]
Oops

**Comment by trivialfis** [comment]
Sorry.  i thought you sorted out stuffs with offline conversations ...

**Comment by trivialfis** [comment]
@CodingCat Do you want me to revert?

**Comment by CodingCat** [comment]
;) one second lag, please help to pull it back for now, I will take a review in weekend

**Comment by trivialfis** [comment]
https://github.com/dmlc/xgboost/pull/4965

**Comment by chenqin** [comment]
I got same impression.. adding https://github.com/dmlc/xgboost/pull/4966

**Comment by CodingCat** [comment]
> I got same impression.. adding #4966

I am not sure how a quick going through lasting no more than 10 mins would create such a impression...and the quality of code ranging from style to test coverage here and in the new PR is not anything close to ready to merge

**Pull request merged**